src/llm/model.py

```diff
@@ -0,0 +1,285 @@
+                value=-torch.inf,
+            )
+        attention_scores = torch.softmax(
+            scaled_similarity, dim=-1
+        )  # Careful with the `dim`
+
+        # project the value
+        value = self.W_value(value)
+        outputs = torch.matmul(attention_scores, value)
+
+        return outputs
+
+
+class MultiHeadAttention(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_heads: int,
+    ):
+        super().__init__()
+        dim_k = dim_v = model_dim // num_heads
+        self.projection_heads = list()
+        for _ in range(num_heads):
+            self.projection_heads.append(
+                ScaledDotProductAttention(model_dim=model_dim, dim_k=dim_k, dim_v=dim_v)
+            )
+        self.W_out = nn.Linear(model_dim, model_dim)
+
+    def forward(
+        self,
+        query: Float32[Tensor, "batch seq_len model_dim"],
+        key: Float32[Tensor, "batch seq_len model_dim"],
+        value: Float32[Tensor, "batch seq_len model_dim"],
+        is_causal: bool=False,
+    ) -> Float32[Tensor, "batch seq_len model_dim"]:
+        outputs = list()
+        for head in self.projection_heads:
+            outputs.append(head(query, key, value, is_causal=is_causal))
+
+        outputs = torch.concat(outputs, dim=-1)
+        outputs = self.W_out(outputs)
+        return outputs
+
+
+class EncoderLayer(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_heads: int,
+        expansion_dim: int,
+    ):
+        super().__init__()
+        self.multi_head_attention = MultiHeadAttention(
+            model_dim=model_dim, num_heads=num_heads
+        )
+        self.feed_forward = FeedForward(
+            model_dim=model_dim, expansion_dim=expansion_dim
+        )
+        self.layer_norm1 = nn.LayerNorm(normalized_shape=model_dim)
+        self.layer_norm2 = nn.LayerNorm(normalized_shape=model_dim)
+
+    def forward(
+        self, inputs: Float32[Tensor, "batch enc_seq_len model_dim"]
+    ) -> Float32[Tensor, "batch enc_seq_len model_dim"]:
+        residual = inputs
+
+        x = self.multi_head_attention(query=inputs, key=inputs, value=inputs)
+        x = self.layer_norm1(x) + residual
+
+        residual = x
+
+        x = self.feed_forward(x)
+        x = self.layer_norm2(x) + residual
+        return x
+
+
+class Encoder(nn.Module):
+    def __init__(
+        self,
+        model_dim: int,
+        num_layers: int,
+        num_heads: int,
+        expansion_dim: int,
+    ):
+        super().__init__()
+        self.encoder_layers = list()
+        for _ in range(num_layers):
+            self.encoder_layers.append(
+                EncoderLayer(
+                    model_dim=model_dim,
+                    num_heads=num_heads,
+                    expansion_dim=expansion_dim,
+                )
+            )
+
+    def forward(
+        self,
+        encoder_inputs: Float32[Tensor, "batch enc_seq_len model_dim"],
```

Comment:
nit: let's call this argument `encoder_input`. I thought about and it feels more readable if we use singular form. I know we are batching but let's ignore it.

we give `Encoder` an `encoder_input` and get out `encoder_outputs` because we are processing through each block and collecting it to be fed to the `Decoder`.

Reply:
This weirdly makes more sense. Changing.